### PR TITLE
Add field range checks in 1.4 -- SIMICS-19194

### PR DESCRIPTION
**Pull request by lwaern-intel**
Design questions still pending, such as whether overlapping ranges should be an error or a warning.

### CODE REVIEW

**Comment by syssimics** [comment]
Verification [#9950418](https://teamcity01-ir.devtools.intel.com/viewLog.html?buildId=9950418): fail

**Comment by syssimics** [comment]
Verification [#9962579](https://teamcity01-ir.devtools.intel.com/viewLog.html?buildId=9962579): fail

**Comment by syssimics** [comment]
Verification [#9963983](https://teamcity01-ir.devtools.intel.com/viewLog.html?buildId=9963983): pass

File: py/dml/structure.py [hunk 1]
```diff
@@ -2199,6 +2200,47 @@ def explode_offsets(reg):
...
+                    raise EBITRN(field, msb, lsb)
+                ranges.append((lsb, msb, indices))
+        return ranges
+
+    # list of quadruples (lsb, msb, DMLObject, indices)
+    all_ranges = []
+    for field in reg.get_recursive_components('field'):
+        all_ranges.extend((lsb, msb, field, indices)
+                          for (lsb, msb, indices) in explode_ranges(field))
+    all_ranges.sort(key=lambda t: t[0])
```
Comment (mandolaerik) [inline py/dml/structure.py:2239]: you could consider `yield` and/or `sorted((lsb, ...) for field in .. for (..) in explode(..))`.

File: test/1.2/misc/devinfo.dml [hunk 2]
```diff
@@ -15,8 +15,8 @@ bank b[j in 0..3][k in 0..1] "bank desc" {
     }
     register r1[2] size 4 @ 18 + $i * 4 {
         parameter documentation = "reg1 doc";
-        field f0[7:3] "f0 desc";
-        field f1[2:1];
+        field f0[9:5] "f0 desc";
+        field f1[4:1];
```
Comment (mandolaerik) [inline test/1.2/misc/devinfo.dml:19]: why not an array here, like in 1.4? 1.2 is forward-compatible with the syntax `field f1[i in 0..1] @ [2*i+2:2*i+1]`

File: py/dml/structure.py [hunk 3]
```diff
@@ -2199,6 +2200,47 @@ def explode_offsets(reg):
...
+                *([mkIntegerLiteral(field.site, i) for i in range(dimsize)]
+                  for dimsize in field.dimsizes[reg.dimensions:])):
+            reg_indices = tuple(StaticIndex(field.site, var)
+                                for var in reg.idxvars())
+            indices = reg_indices + field_indices
+            lsb_expr = param_expr(field, 'lsb', indices)
+            msb_expr = param_expr(field, 'msb', indices)
+            for expr in (lsb_expr, msb_expr):
+                if isinstance(expr, NonValue):
+                    raise expr.exc()
```
Comment (mandolaerik) [inline py/dml/structure.py:2220]: I think this line lacks a test, e.g. a field `@ [undefined:undefined]`. Bonus points if you can trigger DCE (unmappable bank, unmapped register, or register that overrides get/set/read/write).

File: py/dml/structure.py [hunk 4]
```diff
@@ -2199,6 +2200,47 @@ def explode_offsets(reg):
...
+                if expr.constant and not isinstance(expr.value, int):
+                    raise EBTYPE(expr.site, expr.ctype(), typestr)
```
Comment (mandolaerik) [inline py/dml/structure.py:2222]: lacks a test?

File: test/1.2/misc/devinfo.dml (hunk 2 above)
Comment (lwaern-intel) [reply test/1.2/misc/devinfo.dml:19]: I dunno, I didn't write the original code! I'll do it if no issue crops up.

File: py/dml/structure.py (hunk 3 above)
Comment (lwaern-intel) [reply py/dml/structure.py:2220]: I'll add a test, and try to trigger DCE. I'm not sure the latter is possible, though.

File: test/1.2/misc/devinfo.dml (hunk 2 above)
Comment (mandolaerik) [reply test/1.2/misc/devinfo.dml:19]: ah, I misread the diff in 1.4 code (I thought you moved from non-array to array there). Never mind then.

File: py/dml/structure.py (hunk 1 above)
Comment (lwaern-intel) [reply py/dml/structure.py:2239]: Not sure. I'll hum-haw over it. Any of the three choices would be fine in my eyes.

Comment (lwaern-intel) [reply py/dml/structure.py:2239]: I have decided I don't care unless you insist.

File: py/dml/structure.py (hunk 3 above)
Comment (lwaern-intel) [reply py/dml/structure.py:2220]: Tested to see if DCE avoids the error by testing it with an unmappable bank with unmapped register that overrides read and write. No sell.

File: py/dml/structure.py (hunk 1 above)
Comment (mandolaerik) [reply py/dml/structure.py:2239]: All choices are by far over the threshold for acceptable. I insist that the other choices are slightly better, but I don't insist that you care :)

**Review by mandolaerik — APPROVED** [review]

**Comment by syssimics** [comment]
Verification [#9970478](https://teamcity01-ir.devtools.intel.com/viewLog.html?buildId=9970478): pass

**Pull request merged**